programs/compressed-pda/src/lib.rs

```diff
@@ -52,30 +85,312 @@ pub struct InstructionDataTransfer {
+            // Determine the owner index
+            let owner_index = match accounts.iter().position(|&p| p == utxo.owner) {
+                Some(index) => index as u8, // Found in accounts
+                None => match self.pubkey_array.iter().position(|&p| p == utxo.owner) {
+                    Some(index) => (accounts.len() + index) as u8, // Found in accounts
+                    None => {
+                        // Not found, add to pubkey_array and use index
+                        self.pubkey_array.push(utxo.owner);
+                        (accounts.len() + self.pubkey_array.len() - 1) as u8
+                    }
+                },
+            };
+
+            // Add the lamports index
+            let lamports_index = match self.u64_array.iter().position(|&p| p == utxo.lamports) {
+                Some(index) => index as u8, // Found in accounts
+                None => {
+                    // Not found, add to u64_array and use index
+                    self.u64_array.push(utxo.lamports);
+                    (self.u64_array.len() - 1) as u8
+                }
+            };
+
+            // Serialize the UTXO data, if present
+            let data_serializable = utxo.data.as_ref().map(|data| {
+                // This transformation needs to be defined based on how Tlv can be converted to TlvSerializable
+                Tlv::to_serializable_tlv(data, &mut self.pubkey_array, accounts)
+            });
+
+            // Create and add the InUtxoSerializable
+            let in_utxo_serializable = OutUtxoSerializable {
+                owner: owner_index,
+                lamports: lamports_index,
+                data: data_serializable,
+            };
+            self.out_utxo.push(in_utxo_serializable);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, PartialEq)]
+#[account]
+pub struct InUtxoSerializable {
+    pub owner: u8,
+    pub leaf_index: u32,
+    pub lamports: u8,
+    pub data: Option<TlvSerializable>,
+}
+
+// no need to send blinding is computed onchain
+#[derive(Debug, PartialEq)]
+#[account]
+pub struct OutUtxoSerializable {
+    pub owner: u8,
+    pub lamports: u8,
+    pub data: Option<TlvSerializable>,
+}
+
+#[derive(Debug)]
+#[account]
+pub struct OutUtxo {
     pub owner: Pubkey,
-    pub blinding: [u8; 32],
     pub lamports: u64,
-    pub data: Option<TlvData>,
+    pub data: Option<Tlv>,
 }
 
+// blinding we just need to send the leafIndex
 #[derive(Debug)]
 #[account]
-pub struct TokenOutUtxo {
+pub struct Utxo {
     pub owner: Pubkey,
     pub blinding: [u8; 32],
     pub lamports: u64,
-    pub data: Option<TlvData>,
+    pub data: Option<Tlv>,
+}
+
+impl Utxo {
+    pub fn update_blinding(&mut self, merkle_tree_pda: Pubkey, index_of_leaf: usize) -> Result<()> {
+        self.blinding = Poseidon::hashv(&[
+            &hash(merkle_tree_pda.to_bytes().as_slice()).to_bytes()[0..30],
+            index_of_leaf.to_le_bytes().as_slice(),
+        ])
+        .unwrap();
```

Comment:
```suggestion
        ?;
```

It's already a Result.